import sys
import os
import pymysql
from datetime import datetime
import html
from PyQt6.QtWidgets import QApplication, QMainWindow, QMessageBox
from PyQt6 import uic
from dotenv import load_dotenv 
from konlpy.tag import Kkma
from PyQt6.QtCore import QPropertyAnimation, QEasingCurve   # ì• ë‹ˆë©”ì´ì…˜ìš©
from PyQt6.QtCore import QPoint # QPoint ì„í¬íŠ¸ ì¶”ê°€

# Google GenAI ë¼ì´ë¸ŒëŸ¬ë¦¬ ì„í¬íŠ¸
try:
    # 'google-genai' ë¼ì´ë¸ŒëŸ¬ë¦¬ ì‹œë„
    from google import genai
except ImportError:
    print("ğŸš¨ ì˜¤ë¥˜: 'google-genai' ë¼ì´ë¸ŒëŸ¬ë¦¬ë¥¼ ì°¾ì„ ìˆ˜ ì—†ìŠµë‹ˆë‹¤.")
    print("ì„¤ì¹˜í•˜ë ¤ë©´ í„°ë¯¸ë„ì—ì„œ 'pip install google-genai' ëª…ë ¹ì„ ì‹¤í–‰í•˜ì„¸ìš”.")
    sys.exit(1)

# UI íŒŒì¼ ë¡œë“œ
try:
    form_class = uic.loadUiType("mygemini.ui")[0]
except Exception as e:
    app = QApplication(sys.argv)
    QMessageBox.critical(None, "UI íŒŒì¼ ì˜¤ë¥˜", f"UI íŒŒì¼ì„ ì°¾ì„ ìˆ˜ ì—†ìŠµë‹ˆë‹¤.\n\nì—ëŸ¬ ë‚´ìš©: {e}")
    sys.exit()

class GeminiApp(QMainWindow, form_class):
    
    def __init__(self):
        super().__init__()        
        self.setupUi(self)
        
        # [ìˆ˜ì •] QTextBrowser ì„¤ì • ì¶”ê°€
        # QTextBrowserëŠ” ê¸°ë³¸ì ìœ¼ë¡œ ì½ê¸° ì „ìš©ì…ë‹ˆë‹¤.
        # ë§í¬ê°€ í¬í•¨ëœ ë‹µë³€ì´ ì˜¬ ê²½ìš° ë¸Œë¼ìš°ì €ë¡œ ì—´ë¦¬ê²Œ ì„¤ì •í•©ë‹ˆë‹¤.
        try:
            self.answerDisplay.setOpenExternalLinks(True)
        except AttributeError:
            pass # UI íŒŒì¼ì— í•´ë‹¹ ìœ„ì ¯ì´ ì—†ìœ¼ë©´ ë¬´ì‹œ

        # --- [ìˆ˜ì •ëœ ë¶€ë¶„] API í‚¤ ì„¤ì • (NameError í•´ê²°ì„ ìœ„í•´ í•¨ìˆ˜ ë‚´ë¶€ë¡œ ì´ë™) ---
        # 1. .env íŒŒì¼ì—ì„œ í™˜ê²½ ë³€ìˆ˜ë¥¼ ë¶ˆëŸ¬ì˜µë‹ˆë‹¤.
        load_dotenv()
        
        # 2. í™˜ê²½ ë³€ìˆ˜ì—ì„œ GEMINI_API_KEY ê°’ì„ ì½ì–´ì˜µë‹ˆë‹¤
        api_key = os.getenv("GEMINI_API_KEY")
        self.client = None
        
        if not api_key: 
            # API í‚¤ê°€ ì„¤ì •ë˜ì§€ ì•Šì•˜ì„ ê²½ìš° ê²½ê³  í‘œì‹œ
            QMessageBox.critical(
                self, 
                "API í‚¤ ì˜¤ë¥˜", 
                "âš ï¸ API í‚¤ê°€ ì„¤ì •ë˜ì§€ ì•Šì•˜ìŠµë‹ˆë‹¤.\n"
                ".env íŒŒì¼ì— GEMINI_API_KEYê°€ ì˜¬ë°”ë¥´ê²Œ ìˆëŠ”ì§€ í™•ì¸í•´ì£¼ì„¸ìš”."
            )
            # í´ë¼ì´ì–¸íŠ¸ë¥¼ Noneìœ¼ë¡œ ë‘¡ë‹ˆë‹¤.
        else:
            try:
                # í™˜ê²½ ë³€ìˆ˜ì—ì„œ API í‚¤ë¥¼ ì‚¬ìš©í•˜ì—¬ í´ë¼ì´ì–¸íŠ¸ ì´ˆê¸°í™”
                # ëª…ì‹œì ìœ¼ë¡œ api_keyë¥¼ ì „ë‹¬í•˜ëŠ” ê²ƒì´ ì•ˆì „í•©ë‹ˆë‹¤.
                self.client = genai.Client(api_key=api_key)
            except Exception as e:
                # API ì´ˆê¸°í™” ì‹¤íŒ¨ ì‹œ ì²˜ë¦¬
                error_msg = f"Gemini API í´ë¼ì´ì–¸íŠ¸ ì´ˆê¸°í™” ì˜¤ë¥˜: {e}"
                QMessageBox.critical(self, "API ì˜¤ë¥˜", "Gemini API í´ë¼ì´ì–¸íŠ¸ ì´ˆê¸°í™”ì— ì‹¤íŒ¨í–ˆìŠµë‹ˆë‹¤.\n\n" + error_msg)
                print(error_msg)
                self.client = None
            
        # 4. ë²„íŠ¼ í´ë¦­ ì‹œê·¸ë„ ì—°ê²°
        self.btnSent.clicked.connect(self.ask_gemini) 
        self.btnSent.setVisible(False)
        # Enter í‚¤ ì…ë ¥ ì‹œì—ë„ ì‘ë™í•˜ë„ë¡ ì—°ê²°
        self.lineEditMyQuestion.returnPressed.connect(self.ask_gemini)        
        
        # label_2 ì›ë˜ ìœ„ì¹˜ ì €ì¥
        self.label2_origin = self.label_2.pos()

        # ì¢Œìš° í”ë“¤ê¸° ì• ë‹ˆë©”ì´ì…˜
        self.label2_anim = QPropertyAnimation(self.label_2, b"pos")
        self.label2_anim.setDuration(600)   # 0.6ì´ˆ ì™•ë³µ
        self.label2_anim.setLoopCount(-1)   # ë¬´í•œ ë°˜ë³µ
        self.label2_anim.setEasingCurve(QEasingCurve.Type.InOutQuad)

        # í”ë“¤ë¦¼ ë²”ìœ„ ì„¤ì • (ì¢Œìš° 10px)
        x, y = self.label2_origin.x(), self.label2_origin.y()
        self.label2_anim.setStartValue(self.label2_origin)
        self.label2_anim.setKeyValueAt(0.5, self.label2_origin + QPoint(10, 0))
        self.label2_anim.setEndValue(self.label2_origin)
    
    def start_label2_animation(self):
        self.label2_anim.start()

    def stop_label2_animation(self):
        self.label2_anim.stop()
        self.label_2.move(self.label2_origin)   # ìœ„ì¹˜ ì›ìœ„ì¹˜ ë³µê·€

    def ask_gemini(self): 
        # API í´ë¼ì´ì–¸íŠ¸ ì´ˆê¸°í™” ì‹¤íŒ¨ ì‹œ ì²˜ë¦¬
        if not self.client:
            self.answerDisplay.setText("Gemini API í´ë¼ì´ì–¸íŠ¸ê°€ ì´ˆê¸°í™”ë˜ì§€ ì•Šì•˜ìŠµë‹ˆë‹¤. API í‚¤ë¥¼ í™•ì¸í•˜ì„¸ìš”. [ì œë¯¸ë‚˜ì´nh]")
            return      

        question = self.lineEditMyQuestion.text().strip()

        if not question:
            self.answerDisplay.setText("ì§ˆë¬¸ì„ ì…ë ¥í•´ì£¼ì„¸ìš”. [ì œë¯¸ë‚˜ì´nh]")
            return
        
        # ì§ˆë¬¸ ì…ë ¥ì°½ ë¹„ìš°ê¸°
        self.lineEditMyQuestion.clear()
        
        # ë¨¼ì € DBì—ì„œ ê²€ìƒ‰ ì‹œë„
        self.start_label2_animation()  # ì• ë‹ˆë©”ì´ì…˜ ì‹œì‘
        if self.search_mysql(search_text=question) == True:
            self.stop_label2_animation()  # ì• ë‹ˆë©”ì´ì…˜ ì¤‘ì§€ 
            return  # ê²€ìƒ‰ ê²°ê³¼ê°€ ìˆìœ¼ë©´ ìƒˆ ì§ˆë¬¸ ì²˜ë¦¬ ì¤‘ë‹¨


        # ì‘ë‹µ ëŒ€ê¸° ë©”ì‹œì§€ í‘œì‹œ (HTML)
        waiting_html = f"<div>â¡ï¸ ì§ˆë¬¸: <b>{html.escape(question)}</b></div>" \
                   f"<div style='color:gray;'>Geminiê°€ ì‘ë‹µì„ ìƒì„±í•˜ëŠ” ì¤‘ì…ë‹ˆë‹¤... ì ì‹œë§Œ ê¸°ë‹¤ë ¤ì£¼ì„¸ìš”.</div>"
        self.answerDisplay.setPlainText("")             # ë¨¼ì € ì§€ìš°ê³ 
        self.answerDisplay.setHtml(waiting_html)
        QApplication.processEvents() # UI ê°±ì‹  (ë°˜ë“œì‹œ í•„ìš”)

        try:
            # API í˜¸ì¶œ
            response = self.client.models.generate_content(
                model='gemini-2.5-flash',
                contents=question
            )

            # ì‘ë‹µ í‘œì‹œ ë° [ì œë¯¸ë‚˜ì´nh] ì¶”ê°€
            esc_question = html.escape(question).replace('\n', '<br>')
            esc_response = html.escape(response.text).replace('\n', '<br>')
            html_content = (
                f"<div style='color:#1E90FF; font-weight:bold;'>[Gemini ìƒì„± ì‘ë‹µ]</div>"                
                f"<div>â¡ï¸ ì§ˆë¬¸: <b><span style='color:red;'>{esc_question}</span></b></div>"
                f"<hr>"
                f"<div style='color:green; white-space:pre-wrap;'>{esc_response}</div>"
                f"<div style='color:gray; margin-top:8px;'>[ì œë¯¸ë‚˜ì´nh]</div>"
            )

            self.answerDisplay.setPlainText("")                   # ë¨¼ì € ì§€ìš°ê³ 
            self.answerDisplay.setHtml(html_content )   # ìƒˆ ê²°ê³¼ ì¶œë ¥
            
            # (ë‹µë³€ í‘œì‹œ í›„)
            self.save_to_mysql(question, response.text)
            
        except Exception as e:
            # API í˜¸ì¶œ ì¤‘ ì˜ˆì™¸ ì²˜ë¦¬
            error_message = f"API í˜¸ì¶œ ì¤‘ ì˜¤ë¥˜ ë°œìƒ: {e}"
            print(error_message)
            err_html = f"<div>â¡ï¸ ì§ˆë¬¸: <b>{html.escape(question)}</b></div>" \
                       f"<div style='color:red;'>ğŸš¨ ì˜¤ë¥˜: {html.escape(str(error_message))}</div>" \
                       f"<div style='color:gray; margin-top:8px;'>[by geminiNoh]</div>"
            self.answerDisplay.setPlainText("")            # ë¨¼ì € ì§€ìš°ê³ 
            self.answerDisplay.setHtml(err_html)
        finally:
            self.stop_label2_animation()  # ì• ë‹ˆë©”ì´ì…˜ ì¤‘ì§€

    def save_to_mysql(self, question, answer):
        conn = None
        conn2 = None
        try:
            # 1. í˜„ì¬ ì‹œê°„ êµ¬í•˜ê¸°
            current_time = datetime.now().strftime('%Y-%m-%d %H:%M:%S')

            # 2. DB ì—°ê²°
            conn = pymysql.connect( 
                host='bitnmeta2.synology.me',
                user='iyrc',
                passwd='Dodan1004!',
                db='gemini_ai',
                charset='utf8',
                port=3307,  
                cursorclass=pymysql.cursors.DictCursor
            )

            with conn.cursor() as cursor:
                sql = "INSERT INTO chat_history (question, answer, create_at) VALUES (%s, %s, %s)"
                cursor.execute(sql, (question, answer, current_time))
            
            conn.commit()
            print(f"âœ… MySQL ì €ì¥ ì„±ê³µ: {current_time}")
        
        except Exception as e:
            # MySQL Data too long for column -> ì—ëŸ¬ì½”ë“œ 1406 ì²˜ë¦¬
            err_str = str(e)
            print(f"âŒ ë°ì´í„°ë¥¼ ìš”ì•½í•˜ê³  ìˆìŠµë‹ˆë‹¤.: {err_str}")

            is_data_too_long = False
            try:
                if hasattr(e, 'args') and e.args:
                    if isinstance(e.args[0], int) and e.args[0] == 1406:
                        is_data_too_long = True
                if '1406' in err_str or 'Data too long' in err_str:
                    is_data_too_long = True
            except Exception:
                is_data_too_long = False

            if is_data_too_long:
                # ê¸°ì¡´ ì—°ê²° ì•ˆì „í•˜ê²Œ ì¢…ë£Œ
                if conn:
                    try: conn.close()
                    except: pass
                    conn = None

                # ìš”ì•½ ì‹œë„
                summarized = None
                try:
                    if self.client:
                        prompt = ("ì•„ë˜ í…ìŠ¤íŠ¸ë¥¼ í•œêµ­ì–´ë¡œ 500ì ì´ë‚´ë¡œ ìš”ì•½í•´ ì£¼ì„¸ìš”.\n\n" + str(answer))
                        summ_resp = self.client.models.generate_content(
                            model='gemini-2.5-flash',
                            contents=prompt
                        )
                        summarized = summ_resp.text.strip()
                except Exception as se:
                    print(f"ìš”ì•½ ì‹œë„ ì¤‘ ì˜¤ë¥˜: {se}")

                if not summarized:
                    summarized = str(answer)[:500]
                if len(summarized) > 500:
                    summarized = summarized[:500]

                # ì¬ì‹œë„: ìƒˆ ì—°ê²°ë¡œ ì•ˆì „í•˜ê²Œ INSERT
                try:
                    conn2 = pymysql.connect(
                        host='bitnmeta2.synology.me',
                        user='iyrc',
                        passwd='Dodan1004!',
                        db='gemini_ai',
                        charset='utf8',
                        port=3307,
                        cursorclass=pymysql.cursors.DictCursor
                    )
                    with conn2.cursor() as cursor2:
                        sql = "INSERT INTO chat_history (question, answer, create_at) VALUES (%s, %s, %s)"
                        cursor2.execute(sql, (question, summarized, current_time))
                    conn2.commit()
                    print(f"âœ… MySQL ìš”ì•½ ì €ì¥ ì„±ê³µ: {current_time}")
                    
                    # ì‚¬ìš©ìì—ê²Œ ì•Œë¦¼
                    notice_html = f"<div style='color:gray;'>ì›ë¬¸ì´ ê¸¸ì–´ ìš”ì•½(500ì ì´ë‚´)ìœ¼ë¡œ ì €ì¥í–ˆìŠµë‹ˆë‹¤.</div>"
                    try:
                        prev_html = self.answerDisplay.toHtml() # ê¸°ì¡´ ë‚´ìš© ê°€ì ¸ì˜¤ê¸°ê°€ í•„ìš” ì—†ì„ë• ìƒëµ ê°€ëŠ¥
                        self.answerDisplay.setPlainText("")            # ë¨¼ì € ì§€ìš°ê³ 
                        self.answerDisplay.setHtml(prev_html + notice_html)
                    except Exception:
                        self.answerDisplay.append("ì›ë¬¸ì´ ê¸¸ì–´ ìš”ì•½(500ì ì´ë‚´)ìœ¼ë¡œ ì €ì¥í–ˆìŠµë‹ˆë‹¤.")
                        
                except Exception as re:
                    print(f"âŒ ìš”ì•½ ì¬ì €ì¥ ì‹¤íŒ¨: {re}")
        
        finally:
            if conn:
                try: conn.close()
                except: pass
            if conn2:
                try: conn2.close()
                except: pass

    def search_mysql(self, search_text=None):
        """
        DB ê²€ìƒ‰ í•¨ìˆ˜(ëª…ì‚¬ ì¶”ì¶œ ê¸°ë°˜ ë‹¤ì¤‘ ê²€ìƒ‰).
        - search_textê°€ ìˆìœ¼ë©´ ê·¸ê²ƒ ê¸°ë°˜ìœ¼ë¡œ ìˆ˜í–‰.
        - ì—†ìœ¼ë©´ lineEditMyQuestion ë‚´ìš©ìœ¼ë¡œ ê²€ìƒ‰.
        - konlpyë¥¼ ì´ìš©í•´ ëª…ì‚¬ë¥¼ ì¶”ì¶œí•˜ê³ , ê° ëª…ì‚¬ë¥¼ LIKE ê²€ìƒ‰ ì¡°ê±´ìœ¼ë¡œ ì‚¬ìš©í•œë‹¤.
        """
        if search_text is not None:
            text = search_text.strip()
        else:
            text = self.lineEditMyQuestion.text().strip()

        # ---------------------------
        # 1) konlpyë¡œ ëª…ì‚¬ ì¶”ì¶œ
        # ---------------------------
        kkma = Kkma()
        nouns = kkma.nouns(text)

        # ëª…ì‚¬ê°€ ì—†ìœ¼ë©´ ì›ë˜ ë‹¨ì¼ ê²€ìƒ‰ì–´ë¡œ ì‚¬ìš©
        if not nouns:
            nouns = [text]

        # ë„ˆë¬´ ì§§ì€(1ì) ëª…ì‚¬ëŠ” ë³´í†µ ì˜ë¯¸ê°€ ì•½í•˜ë¯€ë¡œ í•„í„°ë§(ì›í•˜ë©´ ì œê±° ê°€ëŠ¥)
        nouns = [n for n in nouns if len(n) > 1]




        # ëª…ì‚¬ê°€ í•˜ë‚˜ë„ ì•ˆ ë‚¨ìœ¼ë©´ ì „ì²´ ë¬¸ì¥ì„ ì‚¬ìš©
        if not nouns:
            nouns = [text]

        try:
            conn = pymysql.connect(
                host='bitnmeta2.synology.me',
                user='iyrc',
                passwd='Dodan1004!',
                db='gemini_ai',
                charset='utf8',
                port=3307,
                cursorclass=pymysql.cursors.DictCursor
            )

            with conn.cursor() as cursor:
                if nouns:
                    # --------------------------------------------
                    # 2) ëª…ì‚¬ë“¤ë¡œ ë‹¤ì¤‘ LIKE ì¡°ê±´ ìƒì„±
                    # --------------------------------------------
                    # question LIKE '%í‚¤ì›Œë“œ%' OR answer LIKE '%í‚¤ì›Œë“œ%'
                    conditions = []
                    params = []

                    for n in nouns:
                        like_n = f"%{n}%"
                        conditions.append("(question LIKE %s OR answer LIKE %s)")
                        params.extend([like_n, like_n])

                    where_clause = " OR ".join(conditions)

                    sql = (
                        "SELECT * "
                        "FROM chat_history "
                        f"WHERE {where_clause} "
                        
                    )

                    cursor.execute(sql, params)

                else:
                    return False

            rows = cursor.fetchall()
            if not rows:
                return False

            # --- 2ì°¨ í•„í„°: ëª…ì‚¬ 80% ì´ìƒ ê²¹ì¹˜ëŠ” rowë§Œ ì„ ë³„ ---
            kkma = Kkma()
            filtered_rows = []

            for row in rows:
                q_text = str(row.get('question', ''))
                a_text = str(row.get('answer', ''))

                row_q_n = kkma.nouns(q_text)
                row_a_n = kkma.nouns(a_text)
                row_nouns = set([n for n in row_q_n + row_a_n if len(n) > 1])

                # êµì§‘í•© ê°œìˆ˜
                overlap = len(row_nouns.intersection(set(nouns)))

                # *** ì¶”ê°€: ê²¹ì¹¨ ë¹„ìœ¨ ê³„ì‚° ***
                if len(nouns) > 0:
                    overlap_ratio = overlap / len(nouns)
                else:
                    overlap_ratio = 0

                # *** ì¡°ê±´: ê²¹ì¹¨ ë¹„ìœ¨ì´ 0.8 ì´ìƒì¼ ë•Œë§Œ ì¸ì • ***
                if overlap_ratio >= 0.8:
                    filtered_rows.append(row)

            
            # í•„í„° í›„ ê²°ê³¼ ì—†ìœ¼ë©´ Gemini í˜¸ì¶œë¡œ ì´ë™
            if not filtered_rows:
                return False

            # ê²°ê³¼ í‘œì‹œ
            lines = []
            lines.append(
                f"<div style='color:#8A2BE2; font-weight:bold;'>[DB ê²€ìƒ‰ ì‘ë‹µ]</div>"
                f"<div style='color:gray;'>ê²€ìƒ‰ì–´: {', '.join(nouns)}</div><hr>"
            )

            for i, row in enumerate(filtered_rows, start=1):
                print("created_at raw value:", row['created_at'], type(row['created_at']))
                created = row.get('created_at', '')              
                # created_atì´ Noneì´ê±°ë‚˜ ë¹ˆ ë¬¸ìì—´ì¸ ê²½ìš° ì²˜ë¦¬
                if created is None or created == '':
                    created = "ì €ì¥ëœ ë‚ ì§œê°€ ì—†ì–´ì„œ " + datetime.now().strftime('%Y-%m-%d %H:%M:%S')
                q = row.get('question', '')
                a = row.get('answer', '')
                esc_q = html.escape(str(q)).replace('\n', '<br>')
                esc_a = html.escape(str(a)).replace('\n', '<br>')

                block = (
                    f"<div style='color:blue; margin-bottom:15px;'>"
                    f"<div><b>{i}. [{html.escape(str(created))}]</b></div>"
                    f"<div><b>Q:</b> <span style='color:red;'>{esc_q}</span></div>"
                    f"<div><b>A:</b> {esc_a}</div>"
                    f"</div>"
                )
                lines.append(block)
            result_html = "".join(lines)
            self.answerDisplay.setPlainText("")            # ë¨¼ì € ì§€ìš°ê³ 
            self.answerDisplay.setHtml(result_html)
            return True

        except Exception as e:
            err = f"DB ê²€ìƒ‰ ì¤‘ ì˜¤ë¥˜ ë°œìƒ: {e}"
            print(err)
            self.answerDisplay.setText(err)
            return False

        finally:
            if 'conn' in locals() and conn:
                conn.close()

if __name__ == "__main__":
    app = QApplication(sys.argv)
    window = GeminiApp()
    window.show()
    sys.exit(app.exec())